### ISSUE TITLE
tensorflow 2.0 keras multi_gpu_model only utilizing one GPU

### ISSUE THREAD

**Issue by mketcha** [tensorflow/tensorflow#30321]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary, pip install
- TensorFlow version (use command below): tensorflow-gpu==2.0.0-beta1
- Python version: 3.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: cudatoolkit 10.0.130, cudnn 7.6.0
- GPU model and memory: 4x NVIDIA GeForce GTX 1080 Ti

**Describe the current behavior**
When using multi_gpu_model (i.e., tf.keras.utils.multi_gpu_model) in tensorflow 2.0 to distribute a job across multiple gpus (4), only one gpu appears to be used. That is when monitoring the GPU usage only one GPU shows substantial dedicated GPU memory usage and GPU utility. 

**Describe the expected behavior**
Each of the 4 GPUs should indicate that memory is being copied to the device and processed.

**Code to reproduce the issue**
While my issue arises with custom code using model.fit_generator, I was able to replicate the issue using model.fit with documentation code provided at https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/utils/multi_gpu_model

```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.applications import Xception
from tensorflow.keras.utils import multi_gpu_model
import numpy as np

num_samples = 1000
height = 224
width = 224
num_classes = 1000

# Instantiate the base model (or "template" model).
# We recommend doing this with under a CPU device scope,
# so that the model's weights are hosted on CPU memory.
# Otherwise they may end up hosted on a GPU, which would
# complicate weight sharing.
with tf.device('/cpu:0'):
    model = Xception(weights=None,
                     input_shape=(height, width, 3),
                     classes=num_classes)

# Replicates the model on 8 GPUs.
# This assumes that your machine has 8 available GPUs.
parallel_model = multi_gpu_model(model, gpus=4) # gpus changed to 4
parallel_model.compile(loss='categorical_crossentropy', optimizer='rmsprop')

# Generate dummy data.
x = np.random.random((num_samples, height, width, 3))
y = np.random.random((num_samples, num_classes))

parallel_model.summary()
# This `fit` call will be distributed on 8 GPUs.
# Since the batch size is 256, each GPU will process 32 samples.
parallel_model.fit(x, y, epochs=20, batch_size=16) #batch_sized changed to 16
```

**Comment by Michael-Turek** [comment]
I have separately encountered the same issue as the OP. The load is distributed across all GPUs as expected with multi_gpu_model in TensorFlow1.13.1, but sits on 1 GPU in 2.0.

Below are screenshots of my nvidia-smi when running the above code with 1.13 and 2.0 respectively.

Have I written custom code: No
OS Platform and Distribution: Ubuntu 18.04
TensorFlow installed from (source or binary): binary
TensorFlow version (use command below): 1.13.1 and 2.0.0-beta1, both gpu
            tensorflow-gpu==1.13.1
            tensorflow-gpu==2.0.0-beta1
Python version: 3.6
CUDA/cuDNN version: CUDA: 10.0, cuDNN: 7.6
GPU model and memory: 4x NVIDIA Tesla M60

Tensorflow 1.13.1:
![tf_1](https://user-images.githubusercontent.com/35641527/60617662-cc05cc80-9d99-11e9-812b-592970181a31.png)

Tensorflow 2.0.0-beta1
![tf_2](https://user-images.githubusercontent.com/35641527/60617668-cdcf9000-9d99-11e9-8514-8f26f546580a.png)

**Comment by srihari-humbarwadi** [comment]
Use distribute strategy for multi gpu training in tf2.0. As of now multi_gpu_model doesnot work when running in eager mode, which is default in tf2.0. Here is the code with necessary changes made
```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.applications import Xception
import numpy as np

num_samples = 1000
height = 224
width = 224
num_classes = 1000

strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    parallel_model = Xception(weights=None,
                     input_shape=(height, width, 3),
                     classes=num_classes)
    parallel_model.compile(loss='categorical_crossentropy', optimizer='rmsprop')

# Generate dummy data.
x = np.random.random((num_samples, height, width, 3))
y = np.random.random((num_samples, num_classes))

parallel_model.summary()
# This `fit` call will be distributed on 8 GPUs.
# Since the batch size is 256, each GPU will process 32 samples.
parallel_model.fit(x, y, epochs=20, batch_size=16) #batch_sized changed to 16
```

**Comment by IuriiZhakun** [comment]
Any plans or advise how to use multiple gpus with fit_generator method?

tensorflow/python/keras/engine/training.py", line 1157, in fit_generator
    raise NotImplementedError('`fit_generator` is not supported for '
NotImplementedError: `fit_generator` is not supported for models compiled with tf.distribute.Strategy.

**Comment by srihari-humbarwadi** [comment]
@IuriiZhakun  you can use the` tf.data.Dataset` api to create a train/validation dataset which can be used with the tf.keras.Model.fit method. `tf.data.Dataset.from_generator` is one the ways to achieve this.

**Comment by IuriiZhakun** [comment]
training gets stuck with multiple gpus at 100% utilization, after yielding few samples.
the same code works fine without MirroredStrategy.
I can I shall try training without keras fit method.
@srihari-humbarwadi, thanks nevertheless!

**Comment by mketcha** [comment]
Thanks for the great suggestions! 

I am also using fit_generator with eager execution, though for me it was just easier to revert back to TF1.14 until multi_gpu_model gets fixed in TF2.0 eager mode

**Comment by mketcha** [comment]
Following @IuriiZhakun 's comment above:

> Any plans or advise how to use multiple gpus with fit_generator method?
> 
> tensorflow/python/keras/engine/training.py", line 1157, in fit_generator
> raise NotImplementedError('fit_generator is not supported for '
> NotImplementedError: fit_generator is not supported for models compiled with tf.distribute.Strategy.

I did not attempt the MirrorStrategy as I am also using fit_generator

**Comment by Path-A** [comment]
Thank you @srihari-humbarwadi. I was at least able to train the model, but I get an unfortunate error at the end about running out of data.

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10.0.18362 Build 18362
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0-beta1
- Python version: 3.6.8
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: 10.0/7.6.1
- GPU model and memory: 2x MSI GeForce RTX 2080 Ti GAMING X TRIO (no NVlink)

I used the same code that you used in your example except that I changed the batch_size to 64 and the number of epochs to 10:

```
# This `fit` call will be distributed on 2 GPUs.
# Since the batch size is 64, each GPU will process 32 samples.
parallel_model.fit(x, y, epochs=10, batch_size=64)
```

It runs perfectly until the final epoch. It always ends with:

```
Epoch 8/10
16/16 [==============================] - 5s 334ms/step - loss: 3590.6503
Epoch 9/10
16/16 [==============================] - 5s 332ms/step - loss: 3597.1092
Epoch 10/10
12/16 [=====================>........] - ETA: 1s - loss: 3603.6067
W0723 14:30:47.582621   232 training_arrays.py:309] Your dataset ran out of data; 
interrupting training. Make sure that your dataset can generate at least `steps_per_epoch * epochs` 
batches (in this case, 160 batches). You may need to use the repeat() function when 
building your dataset.
12/16 [=====================>........] - ETA: 1s - loss: 3603.6067
```

This is only an issue with MirroredStrategy. When I train on a single GPU, there is no issue.

**Comment by ymodak** [comment]
@mketcha You need to switch to ```tf.distribute.Strategy``` in this case. Closing this issue for now. Please reopen if have any further questions using ```tf.distribute.Strategy```. Thanks!

**Comment by Nestak2** [comment]
> @IuriiZhakun you can use the` tf.data.Dataset` api to create a train/validation dataset which can be used with the tf.keras.Model.fit method. `tf.data.Dataset.from_generator` is one the ways to achieve this.

Thank you for the hint, can you provide a few lines of code, if possible connected to your previous code snippet? I am not advanced with tensorflow and can't bring fit_generator to work on distributed strategies.

**Comment by HanneWhitt** [comment]
Using TF 2.0, one solution is to switch off eager execution, which is what prevents multi_gpu_model from working. This can be done as follows after importing tensorflow: 
```
import tensorflow as tf
tf.compat.v1.disable_eager_execution()
```

**Comment by ASLPZHAO** [comment]
When i use tf.distribute.MirroredStrategy() in TF2.0(stable) and there are three GPUs available(gpu_id: 1、2、3), there are four types of  GPUs utilization.
（1）
![1](https://user-images.githubusercontent.com/27215307/68097274-18e8e380-fef1-11e9-8e81-fb79dd361ad7.png)
（2）
![2](https://user-images.githubusercontent.com/27215307/68097281-21d9b500-fef1-11e9-91f1-878f71d28e39.png)
（3）
![3](https://user-images.githubusercontent.com/27215307/68097282-2736ff80-fef1-11e9-8dc4-9f296b4e3747.png)
（4）
![4](https://user-images.githubusercontent.com/27215307/68097284-2c944a00-fef1-11e9-847d-519facae5e07.png)

 **Is this normal?** @srihari-humbarwadi @IuriiZhakun

**Comment by srihari-humbarwadi** [comment]
Looks like your GPU's are not being kept busy by the input pipeline or the batch size is too small. Please follow this guide and go ahead and tune/ profile your data input pipeline!
https://www.tensorflow.org/guide/data_performance

**Comment by asifpatankar** [comment]
As of December 2021, upgrading from tensorflow version 1.13.1 to tensorflow 2.0.0 and with reference from above mentioned solution, the following snippet change worked well.

# Before

```
available_gpus = len(K.tensorflow_backend._get_available_gpus())

if available_gpus > 1:
	model_parallel = multi_gpu_model(model, gpus = available_gpus)
else:
	model_parallel = model
```

# After

```
available_gpus = len(K.tensorflow_backend._get_available_gpus())

if available_gpus > 1:
    strategy = tf.distribute.MirroredStrategy()
    with strategy.scope():
        model_parallel = multi_gpu_model(model, gpus=available_gpus)
else:
    model_parallel = model
```

**Comment by ptdtrinh** [comment]
@asifpatankar: thanks much men! I solved my problem in Tensorflow v2.0....
